rotkehlchen/tests/data_migrations/test_migration_20.py

```diff
@@ -345,3 +344,228 @@ def test_migration_20_edge_case_multiple_swaps(database: DBHandler) -> None:
+            (binance_timestamp_ms, 'E', 'BTC', '0.5', 'trade', (receive_type := HistoryEventSubType.RECEIVE.serialize()), None, event_idx_1),  # noqa: E501
+            (binance_timestamp_ms, 'E', 'BTC', '0.001', 'trade', (fee_type := HistoryEventSubType.FEE.serialize()), None, event_idx_1),  # noqa: E501
+            # another binance trade (ETH/USD)
+            (binance_timestamp_ms, 'E', 'USD', '3000.0', 'trade', spend_type, 'Second trade', (event_idx_2 := '5265d2a4f23913bbe6481e943c3e10d334f02eaa31914578ecdad321cb3c4a67')),  # noqa: E501
+            (binance_timestamp_ms, 'E', 'ETH', '1.5', 'trade', receive_type, None, event_idx_2),
+            (binance_timestamp_ms, 'E', 'ETH', '0.01', 'trade', fee_type, None, event_idx_2),
+            # Sell trade at Kraken
+            (2000000000, 'B', 'ETH', '2.0', 'trade', spend_type, 'Third trade', (event_idx_3 := '667905837df53bc830544b72ebd8f8efd9425a564e393aae718a421e040a1df7')),  # noqa: E501
+            (2000000000, 'B', 'USD', '3600.0', 'trade', receive_type, None, event_idx_3),
+            # Settlement at Coinbase
+            (3000000000, 'G', 'EUR', '8750.00', 'trade', spend_type, 'Settlement', (event_idx_4 := 'a81ce8625134b68de1cef837e0cbb37c8a754ca71124668eb3786202785016f9')),  # noqa: E501
+            (3000000000, 'G', 'BTC', '0.25', 'trade', receive_type, None, event_idx_4),
+            (3000000000, 'G', 'EUR', '5', 'trade', fee_type, None, event_idx_4),
+        ]
+        assert all(expected_events[i] == row for i, row in enumerate(cursor))
+
+    assert len(list(database.user_data_dir.glob('*_pre_recovery_v48.backup'))) == 1  # verify the safety backup was created  # noqa: E501
+
+
+@pytest.mark.parametrize('perform_upgrades_at_unlock', [False])
+@pytest.mark.parametrize('data_migration_version', [19])
+def test_migration_20_recover_lost_trades_no_backup(database: DBHandler) -> None:
+    """Test that recovery handles missing backup gracefully without data loss.
+
+    When events with location hash identifiers are found but no v47 backup exists,
+    the recovery should log an error and leave the database unchanged.
+    """
+    rotki, db_events = MockRotkiForMigrations(database), DBHistoryEvents(database)
+    trigger_event = _create_location_hash_event(Location.BINANCE, TimestampMS(1000000000))
+
+    with database.user_write() as write_cursor:
+        db_events.add_history_events(write_cursor, history=[trigger_event])
+
+    # ensure no v47 backups exist
+    for backup in database.user_data_dir.glob('*_rotkehlchen_db_v47.backup'):
+        backup.unlink()
+
+    with patch('rotkehlchen.data_migrations.manager.MIGRATION_LIST', new=[MIGRATION_LIST[10]]):  # Migration 20 is at index 10  # noqa: E501
+        DataMigrationManager(rotki).maybe_migrate_data()
+
+    with database.conn.read_ctx() as cursor:  # event should remain unchanged
```

Comment:
Check that all contents of the DB are same? Perhaps count of events on top of this?